### ISSUE TITLE
Server selection timed out after 30000 ms

### ISSUE THREAD

**Issue by ajbieber** [Automattic/mongoose#8485]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

To avoid displaying the deprecation warning:

```
DeprecationWarning: current Server Discovery and Monitoring engine is deprecated,
and will be removed in a future version. To use the new Server Discover and Monitoring
engine, pass option { useUnifiedTopology: true } to the MongoClient constructor.
```

we have added in the `useUnifiedTopology` option to our connection options through `mongoose.set('useUnifiedTopology', true);`

This option was added into our code base over 3 months ago (September 26, 2019) without issue. Recently, as of December 20, 2019, we began to encounter an error when `mongoose.connect()` was called, with the following error: **Server selection timed out after 30000 ms**

> _We are currently working to obtain a full stack trace_

Some special points to note which may help in resolving this bug: 

1. In both environments where we are experiencing this issue, the Node servers are deployed in Docker containers, while the MongoDB database is running directly in the environment or on a different server entirely. If necessary, we can release a modified Docker file to help with repro.
2. We have mitigated the issue by commenting out the line `mongoose.set('useUnifiedTopology', true);` as recommended in comments on [issue 8180](https://github.com/Automattic/mongoose/issues/8180). This results in a deprecation warning which is obviously not preferred, but is a good temporary workaround.

**If the current behavior is a bug, please provide the steps to reproduce.**

Unfortunately this bug has been difficult to reproduce. On our production server (MongoDB 3.4.18), this error occurs every single time the server is started up. On our development environment (MongoDB 3.6.12), this error occurs about 80% of the time we startup the server. In our local environments, none of our developers have been able to reproduce this error. In a separate environment, where both the server and database are run in the same container, the issue has also not been able to be reproduced.

Our mongoose connection is displayed below.

```javascript
function connect(url) {
  return new Promise((resolve, reject) => {
    mongoose.set('useFindAndModify', false);
    mongoose.set('useNewUrlParser', true);
    mongoose.set('useCreateIndex', true);
    mongoose.set('useUnifiedTopology', true); // commented out to remedy bug

    mongoose.connect(url, {}, (err) => {
      if (err) {
        return reject(err);
      }
      return resolve();
    });
  });
}
```

**What is the expected behavior?**

A connection should be able to be made between Mongoose and MongoDB without error and without receiving the deprecation warning. 

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**

Node: 10.15.3
Mongoose: 5.8.5
MongoDB: 3.4.18 & 3.6.12 (Experiencing issue in two different environments)

## Related Issues

- [8180](https://github.com/Automattic/mongoose/issues/8180)
- [8156](https://github.com/Automattic/mongoose/issues/8156)
- [8169](https://github.com/Automattic/mongoose/issues/8169)
- [8478](https://github.com/Automattic/mongoose/issues/8478)

**Comment by vkarpov15** [comment]
Some more questions to answer to help debug this:

- Are you connecting to a standalone server, replica set, or sharded cluster?
- Are you connecting to MongoDB Atlas?
- Are you connecting using a `mongodb+srv://` connection string?
- What version of Mongoose were you using on September 26, 2019 when you were able to successfully connect?

**Comment by ajbieber** [comment]
> Some more questions to answer to help debug this:
> 
> * Are you connecting to a standalone server, replica set, or sharded cluster?
> * Are you connecting to MongoDB Atlas?
> * Are you connecting using a `mongodb+srv://` connection string?
> * What version of Mongoose were you using on September 26, 2019 when you were able to successfully connect?

1. We are running all standalone servers
2. We are **not** connected to MongoDB Atlas
3. The connection string we use is in the format: `mongodb://${host}:${port}/${databaseName}`
4. Back on September 26th, we were using Mongoose v5.7.1

**Comment by ajbieber** [comment]
We have been able to resolve the issue on our end and can confirm **there is no bug**.

After further debugging, we were able to retrieve the full stack trace of the error, and discovered the error object has a `reason` field, which contained much more useful information which led us resolving the issues. Both our prod and dev environments had different reasons, but still had the same error printed to the console `Server selection timed out after 30000 ms`.

One thing I might recommend is displaying that reason along with the main Mongoose error message. The timeout error message is far too generic, and doesn't provide enough detail in m opinion.
